### PR TITLE
SAME70 work in progress

### PR DESCRIPTION
**Pull request by tstellanova**
In progress updates for atsame70q21b (mcu on the "ATSAME70 Xplained" eval board I have)

Will create a test app and try uart on the board before marking ready for review. Any feedback welcome in the meanwhile.

### CODE REVIEW

**Comment by michalfita** [comment]
@tstellanova thank for taking part in this, your efforts are really appreciated as I don't have time to work on this.

You mobilized me to focus a bit on this. So, first things first:

- There is no way I can let 863 files of PR to go in, not even telling about reviewing
- If you need SVD and PAC crates for SAME70 chips, let me handle this somewhere this week
- Then you can focus on tidying up your PR just to HAL stuff

**Comment by tstellanova** [comment]
Thank you Michal, I’m happy to submit whatever format/files are helpful.  I’m starting a new project first on atsame70 and then moving it to atsamv71.

File: hal/Cargo.toml [hunk 1]
```diff
@@ -24,6 +24,7 @@ edition = "2018"
 cortex-m = "0.6.0"
 embedded-hal = "0.2.3"
 nb = "0.1.2"
+atsame70q21b  = { version = "0.0.1", path = "../pac/atsame70q21b", optional = true }
```
Comment (michalfita) [inline hal/Cargo.toml:27]: Let's don't add one type for a family at time, add all from SAME70 series.
Then, let's land new support at the end.
This piece of work is already on `development` branch.

File: hal/Cargo.toml [hunk 2]
```diff
@@ -45,6 +46,7 @@ atsams70q21b = { version = "0.0.1", path = "../pac/atsams70q21b", optional = tru
 
 [features]
 device-selected = []
+same70q21  = ["atsame70q21b", "device-selected"]
```
Comment (michalfita) [inline hal/Cargo.toml:49]: Ditto.

File: pac/atsame70q21b/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "atsame70q21b"
```
Comment (michalfita) [inline pac/atsame70q21b/Cargo.toml:2]: I'm not keen to add each particular chip variant separately, rather for the whole family.

**Review by michalfita — COMMENTED** [review]
Rebase your PR. I added foundations for whole SAME70 family in revision B to the `development`, so you can start real work on adding useful guts into the HAL.

My preference would be adding new support at the end, not keeping the alphabetical order.

**Comment by tstellanova** [comment]
It looks like the bulk of what I submitted has been moved to PAC already (along with additional SAME7 family device support). To move forward with the HAL, would it make sense to move the HAL to a separate repo? I'm thinking of how eg stm32-rs splits the pac and hal crates (eg [pacs](https://github.com/stm32-rs/stm32-rs) and one [hal](https://github.com/stm32-rs/stm32f4xx-hal)

**Comment by michalfita** [comment]
STM32 crates are more advanced already. For now it's easier to use for me when this is together, at least unless there are more people interested in this family of MCUs and we could turn these into organisation. At the moment to test the build I need whole thing + extra code I have for my own project which isn't open source.

In the long term I agree both types have to be split into separate repositories. But then would be nice to configure various automation and CI as well. Interest in these crates was non-existent until you came with this PR so I didn't pay that much attention to it.

Let's keep the discussion going. We can move it to gitter or start channel on Matrix like atsamd-rs did.

I'm willing to do anything for the best of this project as long it doesn't mean too much fragmentation.

**Comment by michalfita** [comment]
@tstellanova FYI I have a file called `framework.rs` which is intended to mimic some functions from Atmel Software Framework, it has routines to enable UARTs in PMC, enable some PIO peripherals, and initialize system. This is the foundation for what I wanted to call `atsamx7x-framework` crate and use it for HAL implementation. Another file I have some extra initialisation routines for USARTs, that aren't normally part of HAL, but could be moved into framework. I enclosed them if you're interested.
[framework.zip](https://github.com/michalfita/atsams70-rust/files/4747065/framework.zip)

**Comment by tstellanova** [comment]
Thanks for the framework ref, I'll take a look.  The HAL split suggestion arose because I had a little trouble accessing the hal and pac separately from a consuming project's `Cargo.toml` (I typically use `hal_crate = { git="repo_url" }` to access these while under development, prior to crates.io publication.)
Unfortunately the project that was going to use atsame70 has moved to another M7 processor; however, I've got the dev board and I'm still interested in helping move the HAL forward.

**Comment by michalfita** [comment]
@tstellanova Out of curiosity, can I ask what CPU has been chosen? SAMV, SAME, SAMS aren't very popular among community developers.

I'm thinking about separation, but that makes sense only if we have more people on board and do that as organisation.

**Pull request closed without merging**